<launch>
  <!-- Image Walk arguments -->
  <arg name="size" default="240x240" />
  <arg name="walk_scale" default="1" />
  <arg name="test_image" />
  <arg name="image_topic" default="/camera/image_raw" />
  <arg name="depth_image_topic" default="/camera/depth" />
  <arg name="depth_cloud_topic" default="/camera/points" />
  <arg name="speed" default="60" />
  <arg name="fps" default="4" />
  <arg name="height" default="120" />
  <arg name="overlap" default="0" />
  <arg name="pixel_scale" default="0.01" />
  <arg name="transform_topic" default="/robot_tf" />
  <arg name="camera_frame" default="base_link" />

  <!-- Vision arguments -->
  <arg name="scale" default="1.0" />
  <arg name="rate" default="4" />
  <arg name="hue" default="true" />
  <arg name="intensity" default="true" />
  <arg name="texton" default="true" />
  <arg name="orb" default="true" />
  <arg name="surf" default="false" />
  <arg name="use_pc" default="true" />
  
  <!-- Topic model arguments -->
  <arg name="K" default="100" />
  <arg name="V" default="16180" />
  <arg name="alpha" default="0.1" />
  <arg name="beta" default="1.0" />
  <arg name="gamma" default="0.001" />
  <arg name="tau" default="0.5" />
  <arg name="p_refine_last_observation" default="0.5" />
  <arg name="num_threads" default="4" />
  <arg name="cell_space" default="$(arg pixel_scale)" />
  <arg name="cell_time" default="3600" />
  <arg name="G_time" default="1" />
  <arg name="G_space" default="1" />
  <arg name="polled_refine" default="false" />
  <arg name="words_topic" default="/rost_vision/words" />
  <arg name="min_obs_refine_time" default="200" />
  <arg name="word_obs_queue_size" default="1" />
  <arg name="map_ppx_type" default="global" />

  <!-- Visualization arguments -->
  <!--<arg name="words_topic" default="/rost_vision/words" />-->
  <arg name="topics_topic" default="/rost/topics" />
  <arg name="ppx_topic" default="/rost/cell_perplexity" />
  <arg name="show_topics" default="false" />
  <arg name="show_words" default="true" />
  <arg name="show_perplexity" default="false" />
  <arg name="show_equalized" default="true" />
  
  <!-- Launch image walker -->
  <node pkg="sunshine" type="test.image_walker" name="image_walker" args="$(arg test_image)" >
    <param name="image_topic" type="string" value="$(arg image_topic)"/>
    <param name="size" type="string" value="$(arg size)"/>
    <param name="depth_image_topic" type="string" value="$(arg depth_image_topic)"/>
    <param name="scale" type="double" value="$(arg walk_scale)"/>
    <param name="fps" type="double" value="$(arg fps)" />
    <param name="speed" type="double" value="$(arg speed)" />
    <param name="overlap" type="double" value="$(arg overlap)" />
    <param name="height" type="double" value="$(arg height)" />
    <param name="col_major" type="bool" value="false" />
    <param name="fov" type="double" value="-1" />
    <param name="pixel_scale" type="double" value="$(arg pixel_scale)" />
    <param name="transform_topic" type="string" value="$(arg transform_topic)" />
    <param name="frame_id" type="string" value="$(arg camera_frame)" />
    <param name="depth_cloud_topic" type="string" value="$(arg depth_cloud_topic)" />
    <param name="depth_image_topic" type="string" value="$(arg depth_image_topic)" />
  </node>

  <!--<include file="$(find gscam)/examples/v4l.launch"></include>-->

  <!-- Launch word extractor -->
  <node pkg="sunshine" type="visual_word_extractor" name="rost_vision">
    <!--param name="vocabulary" type="string" value="$(find summarizer)/share/vocab/underwater1k.yml"/-->
    <!--<param name="vocabulary" type="string" value="$(find rost_vision)/data/orb_vocab/barbados2010_5000.yml"/>-->
    <param name="feature_descriptor" type="string" value="ORB"/>
    <param name="scale" type="double" value="$(arg scale)"/>
    <param name="rate" type="double" value="$(arg rate)"/>
    <param name="use_hue" value="$(arg hue)"/>
    <param name="use_intensity" value="$(arg intensity)"/>
    <param name="use_orb" value="$(arg orb)"/>
    <param name="use_surf" value="$(arg surf)"/>
    <param name="use_texton" value="$(arg texton)" />
    <param name="image" value="$(arg image_topic)"/>
    <param name="transform" value="$(arg transform_topic)"/>
    <param name="pc" value="$(arg depth_cloud_topic)" />
    <param name="use_pc" value="$(arg use_pc)" />
  </node>
  

  <!-- Launch topic model -->
  <node pkg="sunshine" type="topic_model" name="rost">
    <param name="K" type="int" value="$(arg K)" />
    <param name="V" type="int" value="$(arg V)" />
    <param name="alpha" type="double" value="$(arg alpha)" />
    <param name="beta" type="double" value="$(arg beta)" />
    <param name="gamma" type="double" value="$(arg gamma)" />
    <param name="tau" type="double" value="$(arg tau)" />
    <param name="p_refine_last_observation" type="double" value="$(arg p_refine_last_observation)" />
    <param name="num_threads" type="int" value="$(arg num_threads)" />
    <param name="cell_space" type="double" value="$(arg cell_space)" />
    <param name="cell_time" type="double" value="$(arg cell_time)" />
    <param name="G_time" type="int" value="$(arg G_time)" />
    <param name="G_space" type="int" value="$(arg G_space)" />
    <param name="polled_refine" value="$(arg polled_refine)" />
    <param name="words_topic" value="$(arg words_topic)" />
    <param name="min_obs_refine_time" value="$(arg min_obs_refine_time)" />
    <param name="word_obs_queue_size" value="$(arg word_obs_queue_size)" />
    <param name="publish_map" type="bool" value="true" />
    <param name="map_ppx_type" type="string" value="$(arg map_ppx_type)" />
  </node>

  <!--
  <node pkg="sunshine" type="visualize" name="visualizer">
    <param name="words_topic" type="string" value="$(arg words_topic)" />
    <param name="topics_topic" type="string" value="$(arg topics_topic)" />
    <param name="ppx_topic" type="string" value="$(arg ppx_topic)" />
    <param name="show_topics" type="bool" value="$(arg show_topics)" />
    <param name="show_words" type="bool" value="$(arg show_words)" />
    <param name="show_perplexity" type="bool" value="$(arg show_perplexity)" />
    <param name="show_equalized" type="bool" value="$(arg show_equalized)" />
   </node>
   -->

  <node pkg="sunshine" type="visualize3d" name="visualizer">
    <param name="observation_topic" type="string" value="$(arg topics_topic)" />
    <param name="output_topic" type="string" value="/topic_cloud" />
  </node>
   <node type="rviz" name="rviz" pkg="rviz" args="-d $(find sunshine)/rviz/imwalker.rviz" />
</launch>
